File: 3 cource/2 sem/FaLP/CW/main.pl
power_set([], []).
power_set([H | T1], [H | T2]) :- power_set(T1, T2).
power_set([_ | T1], T2) :- power_set(T1, T2).

c([], 0, []).
c([H | T1], N, [H | T2]) :- N > 0, N1 is N - 1, c(T1, N1, T2).
c([_ | T1], N, T2) :- c(T1, N, T2).

File: 3 cource/2 sem/FaLP/CW/main.hs
power_set [] = [[]]
power_set (h:t) = do
  array_tail ++ map (h:) array_tail
    where array_tail = power_set t

с [] 0 = []
c array n = do
  filter (\x -> length x == n) result
    where result = power_set array

main = do
  print $ power_set [1, 2, 3]
